generic3g/actions/AccumulatorAction.F90

```diff
@@ -0,0 +1,210 @@
+      procedure :: update
+      ! Helpers
+      procedure :: accumulate
+      procedure :: initialized
+      procedure :: clear_accumulator
+      procedure :: clear_fields
+      procedure :: accumulate_R4
+      procedure :: calculate_result
+   end type AccumulatorAction
+
+contains
+
+   logical function initialized(this) result(lval)
+      class(AccumulatorAction), intent(in) :: this
+
+      lval = ESMF_FieldIsCreated(this%accumulation_field) 
+
+   end function initialized
+
+   subroutine clear_accumulator(this, rc)
+      class(AccumulatorAction), intent(inout) :: this
+      integer, optional, intent(out) :: rc
+      integer :: status
+   
+      call this%clear_fields(_RC)
+      _RETURN(_SUCCESS)
+
+   end subroutine clear_accumulator
+
+   subroutine clear_fields(this, rc)
+      class(AccumulatorAction), intent(inout) :: this
+      integer, optional, intent(out) :: rc
+      
+      integer :: status
+      type(ESMF_TypeKind_Flag) :: tk
+      real(kind=ESMF_KIND_R4) :: CLEAR_VALUE_R4 !wdb fixme deleteme
+
+      call ESMF_FieldGet(this%accumulation_field, typekind=tk, _RC)
+      if(tk == ESMF_TYPEKIND_R4) then
+         call FieldSet(this%accumulation_field, this%CLEAR_VALUE_R4, _RC)
+      else
+         _FAIL('Unsupported typekind')
+      end if
+      _RETURN(_SUCCESS)
+
+   end subroutine clear_fields
+
+   subroutine initialize(this, importState, exportState, clock, rc)
+      class(AccumulatorAction), intent(inout) :: this
+      type(ESMF_State) :: importState
+      type(ESMF_State) :: exportState
+      type(ESMF_Clock) :: clock
+      integer, optional, intent(out) :: rc
+
+      integer :: status
+      type(ESMF_Field) :: import_field
+
+      _HERE
+      call get_field(importState, import_field, _RC)
+      _HERE
+      if(this%initialized()) then
+         call ESMF_FieldDestroy(this%accumulation_field, _RC)
+      end if
+      this%accumulation_field = ESMF_FieldCreate(import_field, _RC)
+      _HERE
+      call this%clear_accumulator(_RC)
+      _HERE
+      _RETURN(_SUCCESS)
+
+   end subroutine initialize
+
+   subroutine update(this, importState, exportState, clock, rc)
+      class(AccumulatorAction), intent(inout) :: this
+      type(ESMF_State) :: importState
+      type(ESMF_State) :: exportState
+      type(ESMF_Clock) :: clock
+      integer, optional, intent(out) :: rc
+
+      integer :: status
+      type(ESMF_Field) :: export_field
+      
+      _ASSERT(this%initialized(), 'Accumulator has not been initialized.')
+      if(.not. this%update_calculated) then
+         call this%calculate_result(_RC)
+         call FieldCopy(this%accumulation_field, this%result_field, _RC)
+      end if
+      call get_field(exportState, export_field, _RC)
+      export_field = this%result_field !wdb fixme deleteme Does this need to be a copy?
+
+      call this%clear_accumulator(_RC)
+      _RETURN(_SUCCESS)
+
+   end subroutine update
+
+   subroutine calculate_result(this, rc)
```

Comment:
This seems to be a poor name.    The purpose seems to be about setting a flag to indicate that things are updated.  So maybe something like `set_updated()`  and then for the flag itself something like `is_current` might be clearer.  But I've never been happy with any choices for this type of flag elsewhere.

Reply:
Absorbed into update with update overloaded for MeanAccumulator.F90 since it has additional calculations. No longer necessary to have a separate calculate_result.